arangod/IResearch/IResearchLink.cpp

```diff
@@ -1000,211 +1007,112 @@ arangodb::Result IResearchLink::initDataStore(InitCallback const& initCallback,
-
-  irs::directory_reader recovery_reader;
-
-  if (pathExists) {
-    try {
-      recovery_reader = irs::directory_reader::open(*(_dataStore._directory));
-    } catch (irs::index_not_found const&) {
-      // ignore
-    }
-  }
-
-  // if this is an existing datastore then ensure that it has a valid
-  // '.checkpoint' file for the last state of the data store
-  // if it's missing them probably the WAL tail was lost
-  if (recovery_reader) {
-    irs::index_file_refs::ref_t ref;
-
-    // find the latest segment state with a checkpoint file
-    for(;;) {
-      auto& filename = recovery_reader.meta().filename; // segment state filename
-      auto checkpointFile = // checkpoint filename
-        filename + std::string(IRESEARCH_CHECKPOINT_SUFFIX);
-
-      ref = irs::directory_utils::reference( // create a reference
-        *(_dataStore._directory), checkpointFile, false // args
-      );
-
-      if (ref) {
-        break; // found checkpoint file for latest state
-      }
-
-      auto src = _dataStore._path;
-      auto& srcFilename = filename;
-      auto dst = src;
-      auto dstFilename = filename + std::string(IRESEARCH_BACKUP_SUFFIX);
-
-      src /= srcFilename;
-      dst /= dstFilename;
-
-      // move segment state file without a matching checkpint out of the way
-      if (!src.rename(dst)) {
-        return arangodb::Result( // result
-          TRI_ERROR_ARANGO_ILLEGAL_STATE, // code
-          std::string("failed rename the latest data store state file for arangosearch link '") + std::to_string(id()) + "', source '" + srcFilename + "' destination '" + dstFilename + "' in path: " + _dataStore._path.utf8()
-        );
-      }
-
-      try {
-        recovery_reader.reset(); // unset to allow for checking for success below
-        recovery_reader = irs::directory_reader::open(*(_dataStore._directory)); // retry opening
-      } catch (irs::index_not_found const&) {
-        // ignore
-      }
-
-      if (!recovery_reader) {
-        return arangodb::Result( // result
-          TRI_ERROR_ARANGO_ILLEGAL_STATE, // code
-          std::string("failed to find checkpoint file matching the latest data store state for arangosearch link '") + std::to_string(id()) + "', expecting file '" + checkpointFile + "' in path: " + _dataStore._path.utf8()
-        );
-      }
+  switch (_engine->recoveryState()) {
+    case RecoveryState::BEFORE: // link is being opened before recovery
+    case RecoveryState::DONE: { // link is being created after recovery
+      _dataStore._inRecovery = true; // will be adjusted in post-recovery callback
+      _recoveryTick = _engine->recoveryTick();
+      break;
     }
 
-    auto& checkpointFile = *ref; // ref non-null ensured by above loop
-    auto in = _dataStore._directory->open( // open checkpoint file
-      checkpointFile, irs::IOAdvice::NORMAL // args, use 'NORMAL' since the file could be empty
-    );
+    case RecoveryState::IN_PROGRESS: { // link is being created during recovery
```

Comment:
question: Rocksdb does not re-create indexes during a recovery, indexes that were only partially build before a crash are just removed. Is that consistent with that comment ?

Reply:
the comment is written with the assumption `IResearchLink == Index` and states that if recovery stepped on link creation marker, link will be created based on the data in restored collection(s).